### ISSUE TITLE
Consider reducing Constants.MAX_ROW_DEGREE from 22 to 8

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9809]
jOOQ 3.0's main new feature were the `Record[N]` types, which provide record level type safety for up to record degrees of 22. 22 was chosen arbitrarily, because Scala knows tuples up to degree 22. C# limits their built-in Tuple types to degrees of 8: https://docs.microsoft.com/en-us/dotnet/api/system.tuple-8?view=netframework-4.8 (where the 8th value can be a nested tuple).

A quick check when removing all the types and methods between degrees 9 and 22 yields the following compilation times for the jOOQ core library:

**With degree 22**

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:57 min
[INFO] Finished at: 2020-02-07T11:04:04+01:00
[INFO] ------------------------------------------------------------------------
```

**With degree 8**

```
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:09 min
[INFO] Finished at: 2020-02-07T11:12:41+01:00
[INFO] ------------------------------------------------------------------------
```

Another obvious gain can be achieved when implementing https://github.com/jOOQ/jOOQ/issues/4695#issuecomment-588966023, which is far less backwards incompatible.

There may be other, more significant benefits, e.g. when compiling client code or working with IDEs. A first, quick check did not yield any significant results. A more thorough investigation will follow.

This is obviously a very significant backwards incompatible change!

**Comment by lukaseder** [comment]
Related idea: `jOOQ-lite`: https://github.com/jOOQ/jOOQ/issues/8928

**Comment by lukaseder** [comment]
Another obvious gain can be achieved when implementing https://github.com/jOOQ/jOOQ/issues/4695#issuecomment-588966023, which is far less backwards incompatible.

**Comment by lukaseder** [comment]
@JanecekPetr: You liked this. I'm curious: What would you gain from this improvement?

**Comment by JanecekPetr** [comment]
> @JanecekPetr: You liked this. I'm curious: What would you gain from this improvement?

Hopefully IDE code completion perf improvements. It got better from 2017, but there's still a bad delay when eclipse is computing its code completion in a jOOQ-heavy class, and subjectively this began with large tables being worked on. I never took any thread dump / profile to see what's going on. Problem is, of course, that we're inserting into some tables with 19 columns, so having the type-safe thingie end at 8 is a type-safety concern...

So, in short, I'm waiting for the "more thorough investigation will follow" to understand whether it would be worth it, and wanted to be notified.

(By the way, I no longer work in my previous company where we were jOOQ-only. Was nice. I can no longer investigate. Here it's jOOQ-marginally :(.)

**Comment by JanecekPetr** [comment]
> Hopefully IDE code completion perf improvements. [...] bad multi-second delay on the main IDE thread when eclipse is computing its code completion [...]

Oh! I just found https://www.eclipse.org/eclipse/news/4.15/jdt.php#non-blocking-completion which will come out next month. Nevermind then.

The JavaDocs are a little bulky because of all those generics, too, but that's not only a problem of the generics, so decreasing that would help, but not too much. I'd consider some API changes to alleviate that, though. Maybe. Things like moving all formatting methods to `.format()` (returns `Formattable`) instead of implementing `Formattable` directly etc.

Um. 22 is fine, then?

**Comment by lukaseder** [comment]
> and subjectively this began with large tables being worked on.

Have you tried setting:

```xml
<recordsImplementingRecordN>false</recordsImplementingRecordN>
```

> Problem is, of course, that we're inserting into some tables with 19 columns, so having the type-safe thingie end at 8 is a type-safety concern...

Well, everything comes at a price :) You could still use the synthetic `INSERT .. SET` syntax rather than `INSERT .. VALUES` to keep type safety.

> So, in short, I'm waiting for the "more thorough investigation will follow" to understand whether it would be worth it, and wanted to be notified.

Makes sense.

> (By the way, I no longer work in my previous company where we were jOOQ-only. Was nice. I can no longer investigate. Here it's jOOQ-marginally :(.)

I'm sure, the jOOQ "virus" will spread also at your new employer's :)

> Oh! I just found https://www.eclipse.org/eclipse/news/4.15/jdt.php#non-blocking-completion which will come out next month. Nevermind then.

Yes, I'm looking forward to that as well.

**Comment by lukaseder** [comment]
With all the exciting new features depending on `Row[N]` https://github.com/jOOQ/jOOQ/issues/11812, https://github.com/jOOQ/jOOQ/issues/11813 and `Record[N]` https://github.com/jOOQ/jOOQ/issues/11804, it would be unwise to reduce the maximum degree of type safety now.

If someone thinks we have too many types, then the jOOQ-lite distribution (#8928) would be for them, if we implement it.
